p2p/server/server.go

```diff
@@ -243,7 +243,15 @@ func (s *Server) Run(ctx context.Context) error {
 				eg.Wait()
 				return nil
 			}
+			ctx, cancel := context.WithCancel(ctx)
 			eg.Go(func() error {
+				<-ctx.Done()
+				// deadlineAdjuster.Close() is safe to call multiple times
+				req.stream.Close()
+				return nil
+			})
+			eg.Go(func() error {
+				defer cancel()
```

Comment:
I think this does the same without needing to spawn a dedicated go routine?

```suggestion
			eg.Go(func() error {
				defer req.stream.Close()
```

Reply:
No, there's a critical difference: if the context is canceled while the handler is still executing (e.g. the server is stopped in a test by means of canceling the context passed to its `Run` method), a blocking I/O operation may still hang and the handler will never exit, also this `defer` will never be executed. In my implementation, the stream is closed immediately after the context is canceled, which may happen not only due `defer cancel()`, but also b/c the context passed to `Run` is canceled; closing the stream interrupts any current blocking operations